File: eprofiler-core/src/main/java/cz/encircled/eprofiler/asm/ProfilerMethodAdapter.java
package cz.encircled.eprofiler.asm;

import cz.encircled.eprofiler.MethodEnd;
import cz.encircled.eprofiler.MethodState;
import cz.encircled.eprofiler.Profiler;
import cz.encircled.eprofiler.Util;
import cz.encircled.eprofiler.registry.MethodDescriptor;
import cz.encircled.eprofiler.registry.MethodRegistry;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.Type;
import org.objectweb.asm.commons.AdviceAdapter;

/**
 * @author Vlad on 23-May-16.
 */
public class ProfilerMethodAdapter extends AdviceAdapter {

    private final MethodDescriptor descriptor;

    private int index;

    protected ProfilerMethodAdapter(MethodVisitor mv, int access, String name, String desc, String className) {
        super(Util.version, mv, access, name, desc);
        descriptor = MethodRegistry.register(name, className, desc);
    }

    @Override
    protected void onMethodEnter() {
        mv.visitLdcInsn(descriptor.id);
        int idIndex = newLocal(Type.LONG_TYPE);
        mv.visitVarInsn(Opcodes.LSTORE, idIndex);

        mv.visitVarInsn(Opcodes.LLOAD, idIndex);
        mv.visitMethodInsn(INVOKESTATIC, Type.getInternalName(Profiler.class), "methodStart",
                Type.getMethodDescriptor(Util.getMethod(Profiler.class, "methodStart")), false);
        index = newLocal(Type.getType(MethodState.class));
        mv.visitVarInsn(Opcodes.ASTORE, index);
    }

    @Override
    protected void onMethodExit(int opcode) {
        mv.visitVarInsn(ALOAD, index);
        mv.visitMethodInsn(INVOKEINTERFACE, Type.getInternalName(MethodEnd.class), "end", Type.getMethodDescriptor(Util.getMethod(MethodEnd.class, "end")),
                true);
    }

}

File: eprofiler-core/src/main/java/cz/encircled/eprofiler/asm/ProfilerMethodAdapter.kt
package cz.encircled.eprofiler.asm

import cz.encircled.eprofiler.MethodState
import cz.encircled.eprofiler.Profiler
import cz.encircled.eprofiler.Util
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes
import org.objectweb.asm.Type
import org.objectweb.asm.commons.AdviceAdapter

/**
 * @author Vlad on 23-May-16.
 */
class ProfilerMethodAdapter constructor(mv: MethodVisitor, access: Int, name: String, desc: String) : AdviceAdapter(Util.version, mv, access, name, desc) {

    private var index: Int = 0

    override fun onMethodEnter() {
        mv.visitMethodInsn(Opcodes.INVOKESTATIC, Type.getInternalName(Profiler::class.java), "methodStart",
                Type.getMethodDescriptor(Util.getMethod(Profiler::class.java, "methodStart")), false)
        index = newLocal(Type.getType(MethodState::class.java))
        mv.visitVarInsn(Opcodes.ASTORE, index)

    }

    override fun onMethodExit(opcode: Int) {
        mv.visitVarInsn(Opcodes.ALOAD, index)
        mv.visitMethodInsn(Opcodes.INVOKEINTERFACE, Type.getInternalName(MethodState::class.java), "end", Type.getMethodDescriptor(Util.getMethod(MethodState::class.java, "end")),
                true)
    }

}
